### DISCUSSION TITLE
Adding undetectable noise to a JPEG while converting from HEIC

### DISCUSSION THREAD

**Discussion by buforded** [ImageMagick/ImageMagick#5228] [Digital Image Processing]
I am converting an HEIC to a JPEG as follows:

`magick convert input.HEIC -quality 50 output.JPG`

... in addition to converting the image to jpeg I would also like to add some small amount of random noise to the image, but not enough to be detected with human eyes.

I thought I could do something like:

`magick convert input.HEIC -quality 50 +noise Random  output.JPG`

... but that just makes the entire image random noise.

The goal here is to change the digital signature of the image and confuse some of the lens and signal processing signatures that the camera produces.  At the same time, I don't want a noisy or poor quality image - so I just want to add a tiny bit of randomness over the entire image.

Would love to do this as a percentage basis - maybe 0.5% all pixels random ?

Thank you.

**Comment by snibgo** [comment]
See https://imagemagick.org/script/command-line-options.php#noise , which says:

> Note that Random will replace the image with noise rather than add noise to the image. Use Uniform, if you wish to add random noise to the image.

**Reply by buforded** [reply]
Thank you - using:

`+noise Uniform -attenuate 1.0`

worked perfectly.  However, I am not sure what I have actually done here as I cannot see any difference in the images.

That's good, but if --attenuate is "maximum noise" I am confused as to why the resulting image is indistinguishable from output without the noise addition.

I was expecting `--attenuate` to look terrible and then have to back off to lower and lower attenuate figures to find something acceptable ...

What am I losing/adding with "maximum noise" that I cannot see with a quick glance ?

Thank you.

**Comment by snibgo** [comment]
When using "magick", "-attenuate" is a setting for "+noise", so it needs to come first. If you put it after "+noise", it will have no effect.

If I add uniform noise to toes.png, and compare the output with the input ...
```
magick toes.png +noise Uniform x.png

magick compare -metric RMSE toes.png x.png NULL:
295.539 (0.00450964)
```
... the RMSE difference is 0.45%.

You asked for:

> add some small amount of random noise to the image, but not enough to be detected with human eyes. ... Would love to do this as a percentage basis - maybe 0.5% all pixels random ?"

So a RMSE of 0.45 seems close. As a rule of thumb, I find that a RMSE of less than 1% that is evenly distributed (EDIT: in an ordinary photograph) is not a visible difference.

We can use "-attenuate" values of greater than 1.0. For example:
```
magick toes.png -attenuate 50.0 +noise Uniform x.png

magick compare -metric RMSE toes.png x.png NULL:
14438.2 (0.220312)
```
22% RMSE is clearly visible:
![x](https://user-images.githubusercontent.com/33812368/173872843-13e3b9e9-1e44-41a5-83fd-a4ea9f0cde9f.png)

I don't know what the documentation means by "maximum noise".

EDIT: For attenuate=1, the value that is added to each pixel channel value (on a scale of 0 to 1) is:
```
rand (-1/32, +1/32)
```

... where rand(a,b) is a random number in the interval [a,b).

I don't know why the factor 1/32 was chosen. But this tells you what attenuate to use if you want to restrict the noise within certain limits.

**Reply by buforded** [reply]
Thank you - this was very useful and immediately actionable.

I wonder if you have any comments, or thoughts, on the addition of noise like this in "anonymizing" a photo and disconnecting it from the particular lens and image processing that was used to produce the original ?

Are there any other imagemagick functions, or transformations, that come to your mind when you think of "obfuscating" a photograph ?

Thanks.

**Reply by Jmuccigr** [reply]
I'm confused by the question.

Converting to a fairly lossy jpeg from a non-lossy HEIC format already introduces "noise", doesn't it? If you're worried about other identifying features in the file, delete the exif data when you do the conversion:

```
mogrify -format jpeg -quality 50 -strip filename.heic
```

**Reply by JohnChristianD** [reply]
Is it normal or a bug for Poisson's attenuate to behave inversely to all the other distributions? I don't really know anything about the math behind the distributions.

**Reply by snibgo** [reply]
This is a deliberate feature. See MagickCore/gem.c function GenerateDifferentialNoise().

**Comment by snibgo** [comment]
As Jmuccigr says.

Obfuscating an image is easy. Strip all metadata, and resize or crop to a single pixel. And reduce the colours to grayscale. And threshold at 50%.
